### PR TITLE
Added Scripting Hooks for controlsconfigdefault.tbl

### PR DESCRIPTION
**Pull request by BMagnu**

### CODE REVIEW

**Comment by BMagnu** [comment]
Converted to draft for addition of locking mechanism for controls

File: code/scripting/api/objs/control_binding.cpp
Comment (asarium) [inline code/scripting/api/objs/control_binding.cpp]: Since the actual field is called `locked` it would be appropriate to call this `isLocked`.

Also, the description makes it seem like this is changing something while it simply returns a value.

**Review by asarium — CHANGES_REQUESTED** [review]

File: code/scripting/api/objs/control_binding.h
Comment (asarium) [inline code/scripting/api/objs/control_binding.h]: Please use paths relative to the `code` directory, not relative paths.

File: code/playerman/playercontrol.cpp
Comment (asarium) [inline code/playerman/playercontrol.cpp]: This should be placed in an API function in the controls code to avoid exposing internal variables globally.

File: code/scripting/api/libs/controls.cpp
Comment (asarium) [inline code/scripting/api/libs/controls.cpp]: Please do not use relative paths.

File: code/controlconfig/controlsconfigcommon.cpp
Comment (asarium) [inline code/controlconfig/controlsconfigcommon.cpp]: We generally do not do include Lua code in other tables since it can lead to some very weird situations and generally is not as flexible.

Instead, the usual way to do this is to expose an API function where a script can register a function to be called when something relevant happens. See the `LuaSEXP` code for some examples.

**Review by asarium — CHANGES_REQUESTED** [review]
Sorry, to leave more comments now. For some reason I did not realize that there was more stuff in here when I did the first review...

**Comment by asarium** [comment]
Wait. This is not even the main repo! I guess that code owner thing still applies here for some reason.

If that review is unwanted you are free to ignore it.

File: code/controlconfig/controlsconfigcommon.cpp
Comment (BMagnu) [reply code/controlconfig/controlsconfigcommon.cpp]: This was done to keep consistent with the Hook: [] +Override: [] syntax used for other types of hooks, as these work semantically identical (with hook being executed every time and the optional override disabling usual in-engine handling).
In addition, you shouldn't really change your hook callback mid-game (just like all the other hooks).
If it is still desired to break that analogy for keeping all lua in sct-tables, I'd change it to a function registerHook(fnc hook, bool enabledByDefault = false, bool isOverride = false).
Also, while this is not the main repo, since it is going to be part of the control overhaul, it needs to conform with main-repo standarts anyways.

Comment (asarium) [reply code/controlconfig/controlsconfigcommon.cpp]: A big part about not having Lua code in tables is that it basically forces a script to keep some global data or else it will not be possible to interact between the "main" script (if such a thing exists) and the controls code.

The proposed signature you posted looks good to me at least. If you ware worried about changes mid-mission then it is perfectly acceptable to just throw an error when that is done in any other place than the `On Game Init` hook.

File: code/controlconfig/controlsconfig.cpp
Comment (z64555) [inline code/controlconfig/controlsconfig.cpp]: Could this section be refactored to reduce the different calls to control_fun_lua?

File: code/controlconfig/controlsconfig.cpp
Comment (z64555) [inline code/controlconfig/controlsconfig.cpp]: Can this value be put in the Config_item? maybe repurpose `int used `?

File: code/controlconfig/controlsconfig.cpp
Comment (z64555) [inline code/controlconfig/controlsconfig.cpp]: Trigger and Continuous may be removed, as this switch is within an axis-only function

File: code/controlconfig/controlsconfig.cpp
Comment (z64555) [inline code/controlconfig/controlsconfig.cpp]: Might be good to stuff `action + JOY_AXIS_BEGIN` into a variable at the start of the for loop, since its used in two places.  This would help keep consistency.

File: code/controlconfig/controlsconfig.h
Comment (z64555) [inline code/controlconfig/controlsconfig.h]: Please doxy these two members using the `//!<` escape

File: code/controlconfig/controlsconfigcommon.cpp
Comment (z64555) [inline code/controlconfig/controlsconfigcommon.cpp:82]: man, you really like maps. :P

File: code/controlconfig/controlsconfigcommon.cpp
Comment (z64555) [inline code/controlconfig/controlsconfigcommon.cpp]: Possible memory hog here.  I'm not sure how the implementation of bool maps are these days, but I suspect at worst-case it'll be an int per bool.  If that does happen to be the case then something like a bitset or bitfield should be used.

File: code/controlconfig/controlsconfigcommon.cpp
Comment (z64555) [inline code/controlconfig/controlsconfigcommon.cpp]: Maybe const these, since they're an alias of member that should not change during the scope of control_run_lua

File: code/controlconfig/controlsconfigcommon.cpp
Comment (z64555) [inline code/controlconfig/controlsconfigcommon.cpp:1672]: This needs to duplicate the behavior of `+Disabled` since this is an override section. Meaning, if `+Locked` is not present, it is assumed the item is unlocked.

File: code/scripting/api/objs/control_binding.h
Comment (z64555) [inline code/scripting/api/objs/control_binding.h:16]: Please be sure to doxy everything here, as needed.

**Review by z64555 — CHANGES_REQUESTED** [review]

File: code/controlconfig/controlsconfig.cpp
Comment (z64555) [reply code/controlconfig/controlsconfig.cpp]: I'm actually surprised none of the compilers are complaining about the lack of a default here.

File: code/controlconfig/controlsconfig.cpp
Comment (BMagnu) [reply code/controlconfig/controlsconfig.cpp]: int used is actively used elsewhere, hence I'd be a bit scared of that.
I specifically put it here, since it is only ever required to store the last value of an axis. Rather, only the absolute axes would suffice (as locking those keeps them at their last recorded value), but that's hard to address.
I could add another field in Config_item, but as this is very much not a "part" of the config, only applies to axes, and only a low level value cache, I thought it'd be better to keep it as a separate array.

File: code/controlconfig/controlsconfig.cpp
Comment (BMagnu) [reply code/controlconfig/controlsconfig.cpp]: I'll try

File: code/controlconfig/controlsconfig.cpp
Comment (BMagnu) [reply code/controlconfig/controlsconfig.cpp]: > I'm actually surprised none of the compilers are complaining about the lack of a default here.

Weird behaviour cause it's a switch on an enum, I think. It does however require me to include Trigger and Continuous, as otherwise the compiler will complain about incomplete switches on enums

File: code/controlconfig/controlsconfig.cpp
Comment (BMagnu) [reply code/controlconfig/controlsconfig.cpp]: Okay, will do

File: code/controlconfig/controlsconfig.h
Comment (BMagnu) [reply code/controlconfig/controlsconfig.h]: Okay.

File: code/controlconfig/controlsconfigcommon.cpp
Comment (BMagnu) [reply code/controlconfig/controlsconfigcommon.cpp:82]: I could make it an array, but assuming that most of the time very few controls ever have hooks, I opted for the map

File: code/controlconfig/controlsconfigcommon.cpp
Comment (BMagnu) [reply code/controlconfig/controlsconfigcommon.cpp]: I'll see what I can do

File: code/controlconfig/controlsconfigcommon.cpp
Comment (BMagnu) [reply code/controlconfig/controlsconfigcommon.cpp]: Okay

File: code/controlconfig/controlsconfigcommon.cpp
Comment (BMagnu) [reply code/controlconfig/controlsconfigcommon.cpp:1672]: Okay

File: code/scripting/api/objs/control_binding.h
Comment (BMagnu) [reply code/scripting/api/objs/control_binding.h:16]: okay

File: code/controlconfig/controlsconfig.cpp
Comment (z64555) [reply code/controlconfig/controlsconfig.cpp]: That's probably why the compiler isn't complaining about the missing `default` then, since you defined a case for every member of the enum.  Likewise, if you define a `default` case the compiler should by happy about the cases.

File: code/controlconfig/controlsconfigcommon.cpp
Comment (z64555) [reply code/controlconfig/controlsconfigcommon.cpp:82]: Maps have O(log(n)) time, making them more performant for a larger set of keys vs. a sequential search in an ordered vector.  This is probably splitting hairs, performance wise, but there is a minimum number of keys that the map will perform better vs. the vector, below that minimum the vector does better.  I have to dig out the graphing calculator to show where that crossover point is.

Comment (BMagnu) [reply code/controlconfig/controlsconfigcommon.cpp:82]: I mean, in an area where n is so small that the O(log n) of the Map is slower than the O(n) of the vector, performance is a non-issue anyways. And so long as that is the case, the map implementation is a lot easier to read IMO.

File: code/controlconfig/controlsconfig.cpp
Comment (BMagnu) [reply code/controlconfig/controlsconfig.cpp]: I can add a default statement if you want, but I'd keep the CC_TYPE_AXIS_BTN_POS and CC_TYPE_AXIS_BTN_NEG cases, since they need to be thought about here, depending on implementation.

File: code/controlconfig/controlsconfigcommon.cpp
Comment (z64555) [reply code/controlconfig/controlsconfigcommon.cpp:82]: Ah, I might have confused O(log n) it with O(n log n), sorry.

Anyway, we should strive to seek performant solutions for the controls code, and balance it with memory consumption.  Although it may not seem like a big deal with main memory getting progressively larger, the CPU cache isn't growing as quickly, and since controls are a module that should be kept in the cache every frame, it needs some care with persistent data sizes. (puts away soap box).

Comment (BMagnu) [reply code/controlconfig/controlsconfigcommon.cpp:82]: > Ah, I might have confused O(log n) it with O(n log n), sorry.

map::find is O(log n) (since it essentially binary searches instead of going through every element) and is thus faster than iterating through a vector, unless n is very very small. However, maps can be less efficient at using cache, I admit.
As for memory consumption, map needs slightly more, as there is obviously small overhead for each node in the tree. But that is probably neglectable.

File: code/controlconfig/controlsconfig.cpp
Comment (z64555) [reply code/controlconfig/controlsconfig.cpp]: Yes, both AXIS_BTN cases stay in.  We have an `UNREACHABLE` macro used to break in debug builds, which are meant to stuff in many `default` cases.

Comment (BMagnu) [reply code/controlconfig/controlsconfig.cpp]: I thought about using that, decided against it, since the Assertion actually catches all cases currently leading to where the UNREACHABLE would be

File: code/scripting/api/objs/control_binding.h
Comment (z64555) [reply code/scripting/api/objs/control_binding.h:16]: Doxygen expects the double-star on multi-line comments for documentation purposes.  It will automatically read it as an @brief member, which should be a one-sentance or one-line description.  Long-winded descriptions should be tagged with @details,  example:

```
/**
 * @brief A class wrapping the Control_config's ActionID to Lua.
 * @details Lua can request a control-binding object by name, and will get this class. 
 *   This class, only holding the IoActionId as an identifier, can then be used to access and modify the control binding with the id.
 */
 ```
 
[https://www.doxygen.nl/manual/commands.html](url)

File: code/scripting/api/objs/control_binding.h
Comment (z64555) [inline code/scripting/api/objs/control_binding.h]: typo

File: code/scripting/api/objs/control_binding.h
Comment (z64555) [inline code/scripting/api/objs/control_binding.h]: Style nitpick, "returns" tell what value is returned, here the grammar suggests that it doesn't return anything (or at all) if the ActionId is invalid.
```
/**
 * @returns true if this object holds a valid IoActionId, false otherwise
 */
 ```
 
Loaded values of bools, ints, and enums that have multiple meanings can opt to return the same value, but document each case of why that value is done.  Use your judgment as to whether its cleaner or easier to document them on a single line or multiple lines:
```
/**
 * @f haveEnoughGas()
 * @returns true If we have infinite gas, or
 * @returns true if we have enough gas to get to the target, or
 * @returns false otherwise
 */
 ```
```
 /**
  * @f haveEnoughGas()
  * @returns true if we have infinite gas or enough gas to get there, or
  * @returns false otherwise
  */
 ```

File: code/controlconfig/controlsconfig.cpp
Comment (z64555) [reply code/controlconfig/controlsconfig.cpp]: Ok, so its just a matter of style to use an assert before the switch or `UNREACHABLE` within the default in this case, which I'll leave up to you if you want to change or not. :)

Comment (BMagnu) [reply code/controlconfig/controlsconfig.cpp]: I'd like to keep it as an Assert, since a bad change further up the code could call this function with a different type, making it more of a safety check than a "this code is actually unreachable"-case. Also, it forces to acknowledge the Assert when implementing button axes and not just allows for a stray break in the switch case to solve that.

File: code/controlconfig/controlsconfigcommon.cpp
Comment (z64555) [inline code/controlconfig/controlsconfigcommon.cpp]: This looks good, but it's probably a good idea to wrap the bitset with a class so its users won't need to be careful about which bit they're accessing.  A somewhat moot point now since only these two functions uses it... but there's no telling if it may be repurposed in the future to hold other bool values.

Comment (z64555) [reply code/controlconfig/controlsconfigcommon.cpp]: What that being the case... it may be advantageous to actually add these bools to the CCI's themselves.

Comment (BMagnu) [reply code/controlconfig/controlsconfigcommon.cpp]: This cache is only, and should only ever be used for continuous buttons, which can get queried more than once per evaluation period.
Axes will get queried once a frame and don't need this cache and normal buttons must be able to be called exactly when it's underlying engine code would run (which it currently does).
Hence I'd like to keep this one separate as well, and, if it really needs to be a class, would like to just create an anonymous singleton class here (no other file should also ever really access this)

Comment (z64555) [reply code/controlconfig/controlsconfigcommon.cpp]: As discussed on it was decided to use a bitfield (now wrapped in a singleton class for usage security) to hold lua-specific states of the controls, in contrast to lumping them together with the other bool in the CCI.  This was done on the basis that bool packing within classes is not a C++ standard, and is OS specific, and since we're doing controls routine that's basically firmware level, we need to consider performance and size constraints.

**Review by z64555 — APPROVED** [review]
LGTM, nicely done.

Final call for "Does this work on your machine?" before merge.

**Comment by BMagnu** [comment]
My test case still runs fine for me, even with the new refactors

**Pull request merged**
